chainer/functions/math/batch_l2_norm_squared.py

```diff
@@ -18,29 +21,51 @@ def check_type_forward(self, in_types):
         )
 
     def forward_cpu(self, inputs):
+        self.retain_inputs((0,))
         x = array.as_mat(inputs[0])
         return (x * x).sum(axis=1),
 
     def forward_gpu(self, inputs):
+        self.retain_inputs((0,))
         x = array.as_mat(inputs[0])
         l2normsquared_kernel = cuda.cupy.ReductionKernel(
             'T x', 'T y', 'x * x', 'a + b', 'y = a', '0', 'l2normsquared'
         )
         return l2normsquared_kernel(x, axis=1),
 
-    def backward(self, inputs, gy):
-        x = inputs[0]
-        xp = cuda.get_array_module(x)
-        gy0 = gy[0].reshape(-1, *((1,) * (x.ndim - 1)))
-        if xp is numpy:
-            gx = 2 * x * gy0
-        else:
-            kernel = cuda.elementwise(
-                'T x, T gy', 'T gx', 'gx = 2 * x * gy',
-                'l2normsquared_bwd')
-            gx = kernel(x, gy0)
+    def backward(self, indexes, gy):
+        x = self.get_retained_inputs()
+        return BatchL2NormSquaredGrad().apply((x[0], gy[0]))
+
+
+class BatchL2NormSquaredGrad(function_node.FunctionNode):
+
+    def forward_cpu(self, inputs):
+        self.retain_inputs((0, 1))
+        x, gy0 = inputs
+        gy0 = gy0.reshape(-1, *((1,) * (x.ndim - 1)))
+        gx = 2 * x * gy0
         return gx,
 
+    def forward_gpu(self, inputs):
+        self.retain_inputs((0, 1))
+        x, gy0 = inputs
+        gy0 = gy0.reshape(-1, *((1,) * (x.ndim - 1)))
+        kernel = cuda.elementwise(
+            'T x, T gy', 'T gx', 'gx = 2 * x * gy',
+            'l2normsquared_bwd')
+        gx = kernel(x, gy0)
+        return gx,
+
+    def backward(self, indexes, grad_outputs):
+        x, gy0 = self.get_retained_inputs()
+        gy0 = gy0.reshape(-1, *((1,) * (x.ndim - 1)))
+        gy0 = chainer.functions.broadcast_to(gy0, x.shape)
+        gg2 = 2 * grad_outputs[0]
+        ggx = gg2 * gy0
```

Comment:
The name `gx` would better than `ggx` here because we want the gradient of `x`.

Reply:
Absolutely. Fixed.